pxtlib/sprite-editor/spriteEditor.ts

```diff
@@ -313,73 +314,125 @@ namespace pxtsprite {
+
+            switch (tool) {
+                case PaintTool.Rectangle:
+                    updateIcon(btn, "\uf096", lf("Rectangle"));
+                    break;
+                case PaintTool.Circle:
+                    updateIcon(btn, "\uf10c", lf("Circle"));
+                    break;
+                case PaintTool.Normal:
+                    updateIcon(btn, "\uf040", lf("Pencil"));
+                    break;
+                case PaintTool.Line:
+                    updateIcon(btn, "\uf07e", lf("Line"));
+                    break;
+                default:  // no alternate icon, do not change
+                    return;
             }
-            btn.onClick(() => this.sidebar.setTool(tool));
-            if ((this.activeTool == PaintTool.Circle && tool == PaintTool.Rectangle)
-                    || (this.activeTool == PaintTool.Line && tool == PaintTool.Normal)) {
-                this.setActiveTool(tool);
+
+            btn.onClick(() => {
+                if (tool != PaintTool.Circle && tool != PaintTool.Line) {
+                    this.setIconsToDefault();
+                    this.sidebar.setTool(tool);
+                }
+            });
+
+            function updateIcon(button: TextButton, text: string, title: string) {
+                const shortcut = getPaintToolShortcut(tool);
+
+                button.setText(text);
+                button.title(title);
+                button.shortcut(shortcut);
             }
         }
 
+        setIconsToDefault() {
+            this.switchIconTo(PaintTool.Rectangle);
+            this.switchIconTo(PaintTool.Normal);
+        }
+
         private keyDown = (event: KeyboardEvent) => {
             if (event.keyCode == 16) { // Shift
-                if (!this.shiftDown) {
-                    let btn = (this.sidebar.getButtonForTool(PaintTool.Normal) as TextButton);
-                    btn.setText("\uf07e");
-                    btn.title(lf("Line"));
-                    btn.onClick(() => this.sidebar.setTool(PaintTool.Line));
-                    if (this.activeTool == PaintTool.Normal) {
-                        this.setActiveTool(PaintTool.Line);
-                    }
-                    btn = (this.sidebar.getButtonForTool(PaintTool.Rectangle) as TextButton);
-                    btn.setText("\uf10c");
-                    btn.title(lf("Circle"));
-                    btn.onClick(() => this.sidebar.setTool(PaintTool.Circle));
-                    if (this.activeTool == PaintTool.Rectangle) {
-                        this.setActiveTool(PaintTool.Circle);
-                    }
-                }
                 this.shiftDown = true;
+                this.shiftAction();
             }
 
+            if (event.keyCode == 18 ) { // Alt
+                this.discardEdit();
+                this.paintSurface.setEyedropperMouse(true);
+                this.altDown = true;
+            }
+
+            [
+                PaintTool.Fill,
```

Comment:
Having an array literal as the start of a statement feels weird to me; could we put this in a function or constant?